File: implementations/java/client/src/main/java/com/theshuai/specusclient/peer/PeerEgressSocketDialer.java
package com.theshuai.specusclient.peer;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.StandardProtocolFamily;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.nio.channels.SocketChannel;

/**
 * The real outbound socket.
 *
 * <p>This is the one place the egress touches the network directly, and the only part of the plane
 * that is not driven by the shared vectors: what it produces is a connection, not a value.
 *
 * <p>Each socket is bound, before it connects, to the interface the operating system would have
 * used had this node's own tunnel routes not been there; see {@link PeerEgressSocketBinding}. The
 * sockets are NIO channels rather than {@code java.net.Socket} because a channel's handle can be
 * reached to bind it, and they are opened as IPv4 so the IPv4 option applies to them.
 *
 * <p>On Linux nothing is bound, which is recorded in {@code protocol/spec/peer-egress.md} under
 * current limitations: the forced-deny list refuses this node's own tunnel and virtual interface
 * networks before any connect, which covers the loop, but a Java egress on a node whose tunnel has
 * claimed the default route will send forwarded traffic into the tunnel.
 */
final class PeerEgressSocketDialer implements PeerEgressRuntime.Dialer {

    private final PeerEgressSocketBinder binder;

    PeerEgressSocketDialer(PeerEgressSocketBinder binder) {
        this.binder = binder;
    }

    @Override
    public PeerEgressRuntime.Socket dial(String protocol, String host, int port, long timeoutMs)
            throws IOException {
        InetSocketAddress target = new InetSocketAddress(host, port);
        if (target.isUnresolved()) {
            // Addresses arrive as literals from the flow key, so this means the host was not one.
            throw new IOException("cannot resolve " + host);
        }
        if ("udp".equalsIgnoreCase(protocol)) {
            return new Datagram(target, binder);
        }
        return new Stream(target, timeoutMs, binder);
    }

    /** One TCP connection. */
    private static final class Stream implements PeerEgressRuntime.Socket {
        private final SocketChannel channel;

        Stream(InetSocketAddress target, long timeoutMs, PeerEgressSocketBinder binder) throws IOException {
            channel = SocketChannel.open(StandardProtocolFamily.INET);
            try {
                binder.bind(channel, target);
                channel.socket().connect(target, (int) Math.min(timeoutMs, Integer.MAX_VALUE));
                channel.socket().setTcpNoDelay(true);
            } catch (IOException | RuntimeException failure) {
                closeQuietly();
                throw failure;
            }
        }

        @Override
        public int read(byte[] buffer) throws IOException {
            return channel.read(ByteBuffer.wrap(buffer));
        }

        @Override
        public void write(byte[] data) throws IOException {
            ByteBuffer remaining = ByteBuffer.wrap(data);
            while (remaining.hasRemaining()) {
                channel.write(remaining);
            }
        }

        @Override
        public void closeWrite() {
            try {
                channel.shutdownOutput();
            } catch (IOException alreadyGone) {
                // The consumer's half-close outliving the socket is ordinary.
            }
        }

        @Override
        public void close() {
            closeQuietly();
        }

        private void closeQuietly() {
            try {
                channel.close();
            } catch (IOException ignored) {
                // Teardown races are ordinary; the flow table has already dropped the entry.
            }
        }
    }

    /**
     * One UDP session, connected to the target so a reply from anywhere else is not delivered.
     *
     * <p>Connecting rather than binding is what makes the reply path safe: an unconnected socket
     * would accept a datagram from any source, and the plane would relay it to the consumer as
     * though it came from the address they asked for.
     */
    private static final class Datagram implements PeerEgressRuntime.Socket {
        private final DatagramChannel channel;

        Datagram(InetSocketAddress target, PeerEgressSocketBinder binder) throws IOException {
            channel = DatagramChannel.open(StandardProtocolFamily.INET);
            try {
                binder.bind(channel, target);
                channel.connect(target);
            } catch (IOException | RuntimeException failure) {
                close();
                throw failure instanceof IOException io ? io : new IOException("connect " + target, failure);
            }
        }

        @Override
        public int read(byte[] buffer) throws IOException {
            return channel.read(ByteBuffer.wrap(buffer));
        }

        @Override
        public void write(byte[] data) throws IOException {
            channel.write(ByteBuffer.wrap(data));
        }

        @Override
        public void close() {
            try {
                channel.close();
            } catch (IOException ignored) {
                // Same as the stream.
            }
        }
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerEgressSocketDialer.cs
using System.Net;
using System.Net.Sockets;

namespace Specus.Client.PeerMesh;

/// <summary>
/// The real outbound socket.
/// </summary>
/// <remarks>
/// This is the one place the egress touches the network directly, and the only part of the plane
/// that is not driven by the shared vectors: what it produces is a connection, not a value.
///
/// <para>The egress sends a consumer's traffic to the real internet. If the outbound socket picked
/// up this node's own tunnel route, that traffic would go back into the mesh instead of out, and on
/// a node that is both an egress and a consumer of another egress it would loop. On Windows and
/// macOS the socket is bound, before it connects, to the interface the operating system would have
/// used had those routes not been there; see <see cref="PeerEgressSocketBinding"/>. That binding is
/// not best-effort: a socket that cannot be bound is not opened.</para>
///
/// <para>On Linux the socket is marked so a policy routing rule can steer it to the physical
/// interface regardless of what the tunnel did to the main table. That one is set best-effort: a
/// node without the matching rule, or without permission to set a mark, still works whenever the
/// tunnel did not claim the default route, so failing the connect there would break the common case
/// in order to protect the uncommon one.</para>
/// </remarks>
internal sealed class PeerEgressSocketDialer(PeerEgressSocketBinder binder) : IPeerEgressDialer
{
    /// <summary>
    /// The mark a policy routing rule matches on. The same value the Go client uses, so one
    /// deployment's rule serves either implementation.
    /// </summary>
    private const int EgressSocketMark = 0x5350;

    private const int SolSocket = 1;
    private const int SoMark = 36;

    public IPeerEgressSocket Dial(string protocol, string host, int port, long timeoutMs)
    {
        if (!IPAddress.TryParse(host, out var address))
        {
            // Addresses arrive as literals from the flow key, so this means the host was not one.
            throw new InvalidOperationException($"cannot resolve {host}");
        }
        var target = new IPEndPoint(address, port);
        return protocol.Equals("udp", StringComparison.OrdinalIgnoreCase)
            ? new Datagram(target, EgressSocketMark, binder)
            : new Stream(target, timeoutMs, EgressSocketMark, binder);
    }

    private static void Mark(Socket socket, int mark)
    {
        if (!OperatingSystem.IsLinux())
        {
            // Windows and macOS bind the socket to an interface instead; see the binder.
            return;
        }
        try
        {
            socket.SetRawSocketOption(SolSocket, SoMark, BitConverter.GetBytes(mark));
        }
        catch (SocketException)
        {
            // Best effort; see the class comment.
        }
        catch (PlatformNotSupportedException)
        {
            // Same.
        }
    }

    /// <summary>One TCP connection.</summary>
    private sealed class Stream : IPeerEgressSocket
    {
        private readonly Socket _socket;

        public Stream(IPEndPoint target, long timeoutMs, int mark, PeerEgressSocketBinder binder)
        {
            _socket = new Socket(target.AddressFamily, SocketType.Stream, ProtocolType.Tcp)
            {
                NoDelay = true,
            };
            try
            {
                Mark(_socket, mark);
                binder.Bind(_socket, target);
                _socket.Connect(target);
                _socket.ReceiveTimeout = 0;
                _socket.SendTimeout = (int)Math.Min(timeoutMs, int.MaxValue);
            }
            catch
            {
                _socket.Dispose();
                throw;
            }
        }

        public int Read(byte[] buffer)
        {
            var read = _socket.Receive(buffer);
            // A zero-length receive on a stream socket is the peer's end of stream, which the plane
            // turns into a FIN. Reporting it as zero bytes would spin the reader loop instead.
            return read == 0 ? -1 : read;
        }

        public void Write(byte[] data)
        {
            var sent = 0;
            while (sent < data.Length)
            {
                sent += _socket.Send(data, sent, data.Length - sent, SocketFlags.None);
            }
        }

        public void CloseWrite()
        {
            try
            {
                _socket.Shutdown(SocketShutdown.Send);
            }
            catch (SocketException)
            {
                // The consumer's half-close outliving the socket is ordinary.
            }
            catch (ObjectDisposedException)
            {
                // Same.
            }
        }

        public void Dispose() => _socket.Dispose();
    }

    /// <summary>
    /// One UDP session, connected to the target so a reply from anywhere else is not delivered.
    /// </summary>
    /// <remarks>
    /// Connecting rather than binding is what makes the reply path safe: an unconnected socket
    /// would accept a datagram from any source, and the plane would relay it to the consumer as
    /// though it came from the address they asked for.
    /// </remarks>
    private sealed class Datagram : IPeerEgressSocket
    {
        private readonly Socket _socket;

        public Datagram(IPEndPoint target, int mark, PeerEgressSocketBinder binder)
        {
            _socket = new Socket(target.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
            try
            {
                Mark(_socket, mark);
                binder.Bind(_socket, target);
                _socket.Connect(target);
                _socket.ReceiveTimeout = 0;
            }
            catch
            {
                _socket.Dispose();
                throw;
            }
        }

        public int Read(byte[] buffer) => _socket.Receive(buffer);

        public void Write(byte[] data) => _socket.Send(data);

        public void Dispose() => _socket.Dispose();
    }
}
